core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableLimitRule.java

```diff
@@ -48,6 +48,9 @@
           sort.getCollation(),
           null,
           null);
+      if (input == null) {
```

Comment:
I think this should never happen

Reply:
I understand that now this rule presents a problem if it matches an EnumerableTopNHeapSort, and we should not split it into EnumerableLimit + sortCopy because EnumerableTopNHeapSort is already implementing the Limit logic. So maybe we could add to the rule that `if (sort instanceof EnumerableTopNHeapSort) return;`

Reply:
Re. `input == null`. It's over-defensive code. Remove it.

I think there should be a dedicated rule that converts a `Sort` to an `EnumerableHeapTopNSort`. The rule that creates `EnumerableSort` + `EnumerableLimit` will also fire. Let them compete based on cost.